Enhance order details and payment summary

LevinMedia opened the pull request:
### Changes proposed in this Pull Request:

This pull request makes incremental enhancements to the order details screen. This update also better aligns the order details and payment summary with the way we display the same information in the WooCommerce mobile application.

1. It changes the **cost** of a line item to display it's GROSS amount (before any discounts are applied) This makes it easier to understand how a discount is being applied on a line by line basis.

**Before:**

<img width="868" alt="Screen Shot 2020-01-27 at 7 53 23 PM" src="https://user-images.githubusercontent.com/4500952/73234433-eee53d00-413e-11ea-8983-97e37c96ff19.png">

**After:**

<img width="869" alt="Screen Shot 2020-01-27 at 7 54 02 PM" src="https://user-images.githubusercontent.com/4500952/73234694-17216b80-4140-11ea-8260-f0d5a685a215.png">

2. It adds critical, missing information to the summary section of the order details

- A subtotal of the items
- Fees
- Once the order status becomes "Processing" it displays "Paid by customer" with the amount paid on a new line below the order total

3. It better clarifies new order totals in the case of refunds:

- It displays the total amount refunded below "Paid by customer" along with a new net payment amount

**Before & after**

<img width="2646" alt="Screen Shot 2020-01-27 at 8 17 59 PM" src="https://user-images.githubusercontent.com/4500952/73235255-813b1000-4142-11ea-9d20-421794534136.png">

**Before & after with a refund**

<img width="2644" alt="Screen Shot 2020-01-27 at 9 13 09 PM" src="https://user-images.githubusercontent.com/4500952/73237284-e8a88e00-4149-11ea-89e5-411a446ffe18.png">


### How to test the changes in this Pull Request:

1. Load up this branch
2. Create a new order from scratch - add all the things. Discounts, fees, shipping, taxes
3. Verify all of the lines in the summary are correct, and that their sum matches the order total
4. Pay for the order (the status should update to "Processing")
5. Verify the summary now says "Paid by customer" along with the amount paid (The order total)
6. Refund the order
7. Verify the amount paid minus the refunded amount equals the amount displayed beside "net total"

### Needs testing with additional gateways

I've tested this with the Stripe plugin for WooCommerce, however it would be worth testing other popular gateways to ensure any extra information added by a gateway isn't throwing the summary list out of order. 

In the screen shot below, you'll see that the stripe gateway adds the transaction fee, and the payout amount, however, they're positioned at the bottom of the summary, so the user can simply continue to sum down the column as expected. 

<img width="1434" alt="Screen Shot 2020-01-27 at 9 14 21 PM" src="https://user-images.githubusercontent.com/4500952/73237321-0544c600-414a-11ea-829d-85dbc360e68b.png">


### Other information:

* [X] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Update line item cost to display gross amount. Add items subtotal and fees to order summary, along with "Paid by Customer" once the order status changes to processing

LevinMedia committed:
update order details summary with items subtotal and start to add fees

harvitronix committed:
Abstract fee calculation and expose as method to abstract Object class.

harvitronix committed:
Merge branch 'master' into update/order-details-4

LevinMedia committed:
Merge pull request #25266 from harvitronix/update/order-details-3

Abstract fee calculation and expose as a method on abstract Object class

LevinMedia committed:
rename method to get fee totals

harvitronix committed:
Leave other fee total logic alone, reduce the fee array to get a simple sum

harvitronix committed:
Merge branch 'update/order-details-3' into update/order-details-3

LevinMedia committed:
Merge pull request #25267 from harvitronix/update/order-details-3

Leave other fee total logic alone, reduce the fee array to get a simple sum

LevinMedia committed:
rename fee method again

LevinMedia committed:
update order summary totals

LevinMedia committed:
Rename coupons to discounts in order summary

LevinMedia committed:
display actual item price before discount  in order item cost

LevinMedia committed:
change the order of shipping and fees in order line items

LevinMedia committed:
Add line for net payment post refund

LevinMedia committed:
trying to get net payment to work

LevinMedia committed:
add net payment calculation, and re-order refund item to bottom of rows

LevinMedia committed:
update order to play nice with gateways, add paid by customer when order is paid in full

LevinMedia committed:
update shipping methods

LevinMedia committed:
Change net payment to net total

astralbodies commented:
@LevinMedia If/when this gets merged, can you include the mobile team on the design changes so we will update the apps as well. If the fields/data is in the REST API already to support this, it shouldn't be too hard to implement. cc: @kyleaparker @Garance91540

peterfabian reviewed (dismissed):
Looks good so far, just two items need to be addressed, please see below.

Could you please also rebase this branch to include the tax display fix from 3.9.1?

peterfabian committed:
PHPCS fix.

peterfabian committed:
Merge branch 'master' into update/order-detail-summary

peterfabian commented:
@LevinMedia Would you be able to update this PR or do you prefer us to take it over?

LevinMedia commented:
@peterfabian if ya'll have the bandwith to help me with this one, that would be great 🙌

peterfabian committed:
Replaced deprecated call with a substitute function.

peterfabian committed:
Corrected tax amount displayed in the summary.

peterfabian committed:
PHPCS

Garance91540 commented:
> @LevinMedia If/when this gets merged, can you include the mobile team on the design changes so we will update the apps as well. If the fields/data is in the REST API already to support this, it shouldn't be too hard to implement. cc: @kyleaparker @Garance91540

@astralbodies the only thing that's not currently in our designs are the line items for the "fees". @kyleaparker added it as a new task item for next spring on Monday https://woocommerce.monday.com/boards/241066441/pulses/442781105

LevinMedia commented:
per our team chat this AM, I'm going to add a commit to update the label "discounts" to "Coupon(s)" in the summary section. cc: @Garance91540 @kyleaparker @astralbodies

peterfabian committed:
Renamed Discounts to Coupons.

claudiosanches reviewed (approved):
Works fine, and looks good!

rodrigoprimo committed:
Add basic unit test for WC_Abstract_Order::get_total_fees()

rodrigoprimo committed:
Improve coverage of WC_Abstract_Order::calculate_totals() test

This commits improves the coverage of the
WC_Abstract_Order::calculate_totals() test by including a fee to the
order used to test the method.

rodrigoprimo committed:
Add test to make sure that negative fees never make order total negative

For some reason unknown to me, the following commit added logic to make
sure that negative fees never make the order total negative: https://github.com/woocommerce/woocommerce/commit/804feb93333a8f00d0f93a163c6de58204f31f14

Probably by accident, commit 4326e39250d06002f2b5c736e8c1f19fc982f7d7 changed this logic when `$fees_total += $item->get_total()` was replaced by `$fees_total += $fee_total`:

https://github.com/woocommerce/woocommerce/commit/4326e39250d06002f2b5c736e8c1f19fc982f7d7#diff-9b4164165828b26c4b7aec01c7b17884R1594

This commit adds a failing test that ensures that the old behavior is
tested. This test needs to be fixed before we merge the changes proposed
in PR #25504.

rodrigoprimo commented on includes/admin/meta-boxes/views/html-order-items.php:
```diff
@@ -196,26 +200,62 @@
-		<?php if ( $order->get_total_refunded() ) : ?>
+	<?php if ( $order->get_total_refunded() ) : ?>
+	
+	<table class="wc-order-totals" style="border-top: 1px solid #999; margin-top:12px; padding-top:12px">
```
Minor indentation issue

rodrigoprimo commented on includes/admin/meta-boxes/views/html-order-items.php:
```diff
@@ -196,26 +200,62 @@
+		</tr>
+	<?php endif; ?>
+	
 	</table>
```
Should this closing `</table>` tag be placed inside the if statement above?

peterfabian replied:
Swapped.

rodrigoprimo commented on includes/admin/meta-boxes/views/html-order-items.php:
```diff
@@ -143,15 +143,33 @@
 	<?php endif; ?>
 	<table class="wc-order-totals">
-		<?php if ( 0 < $order->get_total_discount() ) : ?>
+		<?php?>
```
Empty PHP opening and closing tag. Maybe this can be removed or was the original intention to display the "Items Subtotal" only if a given criteria is met in an if statement that was not added?

peterfabian replied:
I think the Items subtotal is added with the intention to always be there, so removing the empty tags.

rodrigoprimo replied:
Yes, I think so. cc @LevinMedia just to double check.

LevinMedia replied:
👍 yep, thank you.

rodrigoprimo commented on includes/admin/meta-boxes/views/html-order-items.php:
```diff
@@ -143,15 +143,33 @@
+					<?php echo wc_price( $order->get_subtotal(), array( 'currency' => $order->get_currency() ) ); // WPCS: XSS ok. ?>
+				</td>
+			</tr>
+		<?php?>
```
Same as above

rodrigoprimo commented on includes/admin/meta-boxes/views/html-order-items.php:
```diff
@@ -196,26 +200,62 @@
+
+	<table class="wc-order-totals">
+		<?php do_action( 'woocommerce_admin_order_totals_after_total', $order->get_id() ); ?>
+
```
Unnecessary extra line?

peterfabian replied:
This line was left intentionally blank :)

rodrigoprimo commented on includes/admin/meta-boxes/views/html-order-items.php:
```diff
@@ -196,26 +200,62 @@
+
+		<table class="wc-order-totals" style="border-top: 1px solid #999; margin-top:12px; padding-top:12px">
+			<tr>
+				<td class="label"><?php esc_html_e( 'Paid By Customer', 'woocommerce' ); ?>:</td>
```
Maybe this was discussed already, but I wonder if adding this "Paid By Customer" line will confuse the store manager when an offline payment gateway is used? I say this because when a customer selects a payment gateway like "Cash on delivery" during checkout, the order status is set to `processing`, but the customer hasn't paid anything yet.

rodrigoprimo replied:
@LevinMedia and @peterfabian, I added commit 44ecb0ec9e509eee91cc2eb69cba892a824271cf to address the issue described above. It basically checks for the presence of the `_date_paid` order meta to decide whether or not to display the "Paid by customer" section. More information on the commit message.

LevinMedia replied:
Thanks @rodrigoprimo !

rodrigoprimo reviewed (changes requested):
Thanks for working on this. The changes proposed in this PR are a great improvement to the order details and payment summary.

I left a few inline comments with questions and suggestions. There is one minor issue that, unless I'm missing something, needs to be fixed before merging this PR. I added a failing test to demonstrate the issue and two other tests to improve the code coverage of the code that is modified here.

peterfabian committed:
Updates to reflect PR feedback.

peterfabian commented on includes/abstracts/abstract-wc-order.php:
```diff
@@ -1695,8 +1709,7 @@ public function calculate_totals( $and_taxes = true ) {
 			}
-
-			$fees_total += $item->get_total();
+			$fees_total += $fee_total;
```
I don't think it was intentional, I reverted it back. Good catch, nice that you added tests as well!

rodrigoprimo committed:
Only display the "Paid by customer" section if the order has been paid

This commit changes the condition in the if the statement that decides
whether or not to display the "Paid by customer" section in the order admin
page. Before it, this section was displayed for all orders that
contained one of the following statuses: 'processing', 'completed' or 'refunded'.

The problem with this approach is that offline payment gateways, like
'Cash on Delivery', set the order status to processing when the order is
placed but before the customer pays for it. So the "Paid by customer"
section was being displayed in some cases where the order was not
actually paid yet creating a confusing experience to store managers.

To fix this, this commit also checks for the presence of the
`_date_paid` order meta. If this order meta is present and the order
status corresponds to one of the three statuses listed above, it means that
the order has been paid and we can display the "Paid by customer"
section. If this order meta is not present, we don't display the "Paid
by customer" section.

rodrigoprimo reviewed (approved):
Thanks for addressing the issues that I mentioned, @peterfabian. This PR looks good to me now. We just need someone else to do a final check on the commit that I added to fix the problem in the "Paid by customer" section that I mentioned here https://github.com/woocommerce/woocommerce/pull/25504#discussion_r374878083

peterfabian reviewed (approved):
This looks good to me now, paid by the customer is added at the right point.
